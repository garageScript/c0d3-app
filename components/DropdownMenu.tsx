import React from 'react'
import DropdownButton from 'react-bootstrap/DropdownButton'
import Dropdown from 'react-bootstrap/Dropdown'
import styles from '../scss/dropDown.module.scss'
import { DropDirection } from 'react-bootstrap/esm/DropdownContext'

//a null item indicates a dropdown divider
export type Item = {
  title: string
  path?: string
  as?: 'a' | 'button'
  onClick?: Function
} | null

type DropDownMenuProps = {
  drop?: DropDirection
  items: Item[]
  title: string
  size?: 'sm' | 'lg' | undefined
  variant?:
    | 'primary'
    | 'secondary'
    | 'success'
    | 'info'
    | 'warning'
    | 'danger'
    | 'none'
  //changes the underlying component CSS base class name
  //https://react-bootstrap.github.io/components/dropdowns/#api
  bsPrefix?: string
}

export const DropdownMenu: React.FC<DropDownMenuProps> = ({
  drop = 'down',
  variant = 'none',
  title,
  size,
  items,
  bsPrefix
}) => {
  const menuItems = items.map((item: Item, itemsIndex: number) =>
    !item ? (
      <Dropdown.Divider key={itemsIndex} />
    ) : (
      <div className="text-center py-2 px-4" key={item.title}>
        <Dropdown.Item
          as={item.as || 'a'}
          key={itemsIndex}
          href={item.path}
          onClick={() => item.onClick && item.onClick(item.title)}
          bsPrefix={bsPrefix}
        >
          {item.title}
        </Dropdown.Item>
      </div>
    )
  )

  return (
    <>
      <div className="d-none d-lg-block">
        <DropdownButton
          title={title}
          variant={variant}
          size={size}
          drop={drop}
          bsPrefix={styles.title}
        >
          {menuItems}
        </DropdownButton>
      </div>
      <div className="d-lg-none">{menuItems}</div>
    </>
  )
}
