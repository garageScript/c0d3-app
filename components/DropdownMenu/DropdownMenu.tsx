import React, { useState } from 'react'
import Dropdown from 'react-bootstrap/Dropdown'
import styles from './dropdownMenu.module.scss'
import { DropDirection } from 'react-bootstrap/esm/DropdownContext'
import { ChevronRightIcon } from '@primer/octicons-react'

//a null item indicates a dropdown divider
export type Item = {
  title: string
  path?: string
  as?: 'a' | 'button'
  onClick?: Function
} | null

type DropDownMenuProps = {
  drop?: DropDirection
  items?: Item[] | null
  title?: string
  //changes the underlying component CSS base class name
  //https://react-bootstrap.github.io/components/dropdowns/#api
  bsPrefix?: string
  customToggle?: {
    Component?: () => JSX.Element
    style?: string
  }
}

const ChevronRight = () => <ChevronRightIcon size={17} />

export const DropdownMenu: React.FC<DropDownMenuProps> = ({
  items,
  title,
  customToggle,
  bsPrefix = ''
}) => {
  const [activeItem, setActiveItem] = useState({ title })

  return (
    <Dropdown bsPrefix={bsPrefix}>
      <Dropdown.Toggle
        bsPrefix={customToggle?.style ? customToggle.style : styles.dropdown}
        id="dropdown-lesson"
        data-testid="dropdown-lesson"
      >
        {customToggle?.Component ? (
          <customToggle.Component />
        ) : (
          <>
            {activeItem.title || 'None'}
            <ChevronRight />
          </>
        )}
      </Dropdown.Toggle>

      <Dropdown.Menu className={styles.dropdown__menu}>
        {items?.map((item, index) =>
          item ? (
            <Dropdown.Item
              key={`${item?.title}-${index}`}
              onClick={() => {
                item?.onClick?.(item)

                setActiveItem({
                  title: item?.title
                })
              }}
            >
              {item?.title}
            </Dropdown.Item>
          ) : (
            <Dropdown.Divider key={index} />
          )
        )}
      </Dropdown.Menu>
    </Dropdown>
  )
}
