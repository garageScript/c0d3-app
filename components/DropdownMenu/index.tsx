export { type Item, DropdownMenu } from './DropdownMenu'
