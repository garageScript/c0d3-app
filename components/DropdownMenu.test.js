import React from 'react'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import { DropdownMenu } from './DropdownMenu'

const dropdownMenuItems = [
  { title: 'Lessons', path: '/admin/lessons', as: 'button' },
  null,
  { title: 'Users', path: '/admin/users', as: 'button' },
  { title: 'Alerts', path: '/admin/alerts', as: 'button' }
]

let testBtnOnClick = ''

describe('MdInput Component', () => {
  test('Should render divider when an item is null', () => {
    const { container, queryByRole } = render(
      <DropdownMenu title="Admin" items={dropdownMenuItems} />
    )
    const div = queryByRole('separator')

    expect(div).toBeTruthy
    expect(container).toMatchSnapshot()
  })

  test('Should change value of testBtnOnClick upon click', () => {
    dropdownMenuItems[0].onClick = val => (testBtnOnClick = val)

    const { container } = render(
      <DropdownMenu title="Admin" items={dropdownMenuItems} />
    )

    const btn = screen.queryByText('Admin')
    fireEvent.click(btn, { button: 1 })

    const lessons = screen.queryAllByText('Lessons')[0]
    fireEvent.click(lessons, { button: 1 })

    waitFor(() => expect(testBtnOnClick).toEqual('Lessons'))
    expect(container).toMatchSnapshot()
  })
})
