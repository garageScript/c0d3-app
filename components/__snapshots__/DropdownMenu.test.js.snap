// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MdInput Component Should change value of testBtnOnClick upon click 1`] = `
<div>
  <div
    class="d-none d-lg-block"
  >
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle title title-none"
        type="button"
      >
        Admin
      </button>
      <div
        aria-labelledby=""
        class="dropdown-menu"
        x-placement="bottom-start"
      >
        <div
          class="text-center px-4"
        >
          <button
            class="dropdown-item"
            data-rr-ui-dropdown-item=""
            href="/admin/lessons"
          >
            Lessons
          </button>
        </div>
        <hr
          class="dropdown-divider"
          role="separator"
        />
        <div
          class="text-center px-4"
        >
          <button
            class="dropdown-item"
            data-rr-ui-dropdown-item=""
            href="/admin/users"
          >
            Users
          </button>
        </div>
        <div
          class="text-center px-4"
        >
          <button
            class="dropdown-item"
            data-rr-ui-dropdown-item=""
            href="/admin/alerts"
          >
            Alerts
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    class="d-lg-none"
  >
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/lessons"
      >
        Lessons
      </button>
    </div>
    <hr
      class="dropdown-divider"
      role="separator"
    />
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/users"
      >
        Users
      </button>
    </div>
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/alerts"
      >
        Alerts
      </button>
    </div>
  </div>
</div>
`;

exports[`MdInput Component Should render divider when an item is null 1`] = `
<div>
  <div
    class="d-none d-lg-block"
  >
    <div
      class="dropdown"
    >
      <button
        aria-expanded="false"
        class="dropdown-toggle title title-none"
        type="button"
      >
        Admin
      </button>
    </div>
  </div>
  <div
    class="d-lg-none"
  >
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/lessons"
      >
        Lessons
      </button>
    </div>
    <hr
      class="dropdown-divider"
      role="separator"
    />
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/users"
      >
        Users
      </button>
    </div>
    <div
      class="text-center px-4"
    >
      <button
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="/admin/alerts"
      >
        Alerts
      </button>
    </div>
  </div>
</div>
`;
